libraries/AP_Frsky_Telem/frsky.cpp

```diff
@@ -0,0 +1,219 @@
+#include <AP_HAL/AP_HAL.h>
+
+#include "frsky.h"
+
+/*
+   State machine to process incoming Frsky SPort bytes
+*/
+bool FRSky::parse_sport_telemetry_data(uint8_t data, sport_parse_state_t &parse_state)
+{
+    switch (parse_state.state) {
+        case ParseState::START:
+            if (data == FRAME_HEAD) {
+                parse_state.state = ParseState::IN_FRAME ;
+                parse_state.telemetry_rx_buffer_count = 0;
+            } else {
+                if (parse_state.telemetry_rx_buffer_count < TELEMETRY_RX_BUFFER_SIZE) {
+                    parse_state.telemetry_rx_buffer[parse_state.telemetry_rx_buffer_count++] = data;
+                }
+                parse_state.state = ParseState::IN_FRAME;
+            }
+            break;
+
+        case ParseState::IN_FRAME:
+            if (data == FRAME_DLE) {
+                parse_state.state = ParseState::XOR; // XOR next byte
+            } else if (data == FRAME_HEAD) {
+                parse_state.state = ParseState::IN_FRAME ;
+                parse_state.telemetry_rx_buffer_count = 0;
+                break;
+            } else if (parse_state.telemetry_rx_buffer_count < TELEMETRY_RX_BUFFER_SIZE) {
+                parse_state.telemetry_rx_buffer[parse_state.telemetry_rx_buffer_count++] = data;
+            }
+            break;
+
+        case ParseState::XOR:
+            if (parse_state.telemetry_rx_buffer_count < TELEMETRY_RX_BUFFER_SIZE) {
+                parse_state.telemetry_rx_buffer[parse_state.telemetry_rx_buffer_count++] = data ^ STUFF_MASK;
+            }
+            parse_state.state = ParseState::IN_FRAME;
+            break;
+
+        case ParseState::IDLE:
+            if (data == FRAME_HEAD) {
+                parse_state.telemetry_rx_buffer_count = 0;
+                parse_state.state = ParseState::START;
+            }
+            break;
+
+    } // switch
+
+    if (parse_state.telemetry_rx_buffer_count >= SPORT_PACKET_SIZE) {
+        parse_state.telemetry_rx_buffer_count = 0;
+        parse_state.state = ParseState::IDLE;
+        return true;
+    }
+    return false;
+}
+
+/*
+ * check sport packet against its crc
+ * skip first byte (START_STOP)
+ * for FrSky SPort Passthrough (OpenTX) protocol (X-receivers)
+ */
+bool FRSky::should_process_sport_packet(const uint8_t *packet)
+{
+    return should_process_sport_packet(packet, nullptr);
+}
+
+bool FRSky::should_process_sport_packet(const uint8_t *packet, const uint8_t *previous_packet)
+{
+#ifdef SPORT_RX_PACKET_DISCARD_DUPLICATE
+    //discard duplicates
+    if (previous_packet != nullptr) {
+        // if they match -> discard this packet
+        if (memcmp(&packet[1], &previous_packet[1], SPORT_PACKET_SIZE-1) == 0) {
```

Comment:
Why isn't the first byte checked here?  If there's any sort of hardware support for memcmp this will be a negative-optimisation if the first byte is expected to be identical.

Reply:
first byte is allowed to differ because same packet can be sent as response to multiple polling bytes (up to 3 in current implementation)